python/src/scipp/compat/dict.py

```diff
@@ -0,0 +1,227 @@
+# SPDX-License-Identifier: GPL-3.0-or-later
+# Copyright (c) 2020 Scipp contributors (https://github.com/scipp)
+# @author Neil Vaytet
+
+from .. import detail
+from .. import _utils as su
+from .._scipp import core as sc
+
+import numpy as np
+from collections import defaultdict
+from itertools import product
+
+
+def to_dict(scipp_obj):
+    """
+    Convert a scipp object (Variable, DataArray or Dataset) to a python dict.
+
+    :param scipp_obj: A Variable, DataArray or Dataset to be converted to a
+                      python dict.
+    :type scipp_obj: Variable, DataArray, or Dataset
+    :return: A dict containing all the information necessary to fully define
+             the supplied scipp object.
+    :rtype: dict
+    """
+    if su.is_variable(scipp_obj):
+        return _variable_to_dict(scipp_obj)
+    elif su.is_data_array(scipp_obj):
+        return _data_array_to_dict(scipp_obj)
+    elif su.is_dataset(scipp_obj):
+        # TODO: This currently duplicates all coordinates that would otherwise
+        # be at the Dataset level onto the individual DataArrays. We are also
+        # manually duplicating all attributes, since these are not carried when
+        # accessing items of a Dataset.
+        out = {}
+        copy_attrs = len(scipp_obj.attrs.keys()) > 0
+        for name, item in scipp_obj.items():
+            out[name] = _data_array_to_dict(item)
+            if copy_attrs:
+                for key, attr in scipp_obj.attrs.items():
+                    out[name]["attrs"][key] = _variable_to_dict(attr)
```

Comment:
Dataset attributes are independent of item attributes. This implementation overwrites/hides the items attributes with those of the dataset, if both are present. Probably not an issue now in practice, and we are planning to change attributes anyway, but something to keep in mind (open an issue?).